Fix Issue 7006 - std.math.pow (integral, integral) crashes on negative exponents

berni44 opened the pull request:
I improved the docs too, including the public unittest.

thewilsonator commented on std/math.d:
```diff
@@ -7954,10 +7969,32 @@ if (isIntegral!(F) && isIntegral!(G))
         }
         break;
     }
+
```
Unneeded.

ibuclaw commented on std/math.d:
```diff
@@ -7912,20 +7912,35 @@ if (isFloatingPoint!(F) && isIntegral!(G))
     typeof(return) p, v = void;
     Unqual!G m = n;
 
+    if (x == -1) return cast(typeof(return)) (m % 2 == 0 ? 1 : -1);
```
`m & 1` is more efficient.

berni44 replied:
Isn't that a job for the optimizer? ;-) Anyway, I changed it.

berni44 commented:
I do not understand, why this always fails on ci/circleci. First I thought it is unrelated, but meanwhile the error was reproduced with each run: std/range/package.test fails. When I run that test locally it succeeds (but needs about 6 minutes). What does ci/circleci actually do?

ahmetsait commented on std/math.d:
```diff
@@ -7912,20 +7912,35 @@ if (isFloatingPoint!(F) && isIntegral!(G))
+ * Returns:
+ *     x raised to the power of n. If n is negative the result is 1 / pow(x, -n),
+ *     which is calculated as integer division with remainder. This may result in
+ *     a division by zero error.
```
I think `1 / pow(x, -n)` supposed to be `1 / pow(x, n)`

thewilsonator commented:
Circle is green now, but `mir/algorithm` still fails consistently.

berni44 commented:
> Circle is green now, but `mir/algorithm` still fails consistently.

Found (and fixed) the problem: `m <= -1` is not a good idea, when `m` is unsigned. I added also a unittest (`assert(pow(0,uint.max) == 0);`) to guard for that problem. (There are two more expressions I put into `static if`, but that is mere for extra security; IMHO they could not break and therefore I could not add a guarding unittest for them.)
